docs/design/ha_master.md

```diff
@@ -0,0 +1,259 @@
+a `ConfigMap` with cluster level configuration, that will keep it. Lack of this map map
+will mean we will use value from the command line flag.
+
+Such cluster-level configuration object is needed also by other components, so
+this `ConfigMap` should be generic enough to serve other needs as well (e.g.
+network related info).
+
+### controller manager, scheduler & cluster autoscaler
+
+Controller manager and scheduler will by default use a lease mechanism to choose an active instance
+among all masters. Only one instance will be performing any operations.
+All other will be waiting in standby mode.
+
+We will use the same configuration in non-replicated mode to simplify deployment scripts.
+
+### add-on manager
+
+All add-on managers will be working independently. Each of them will observe current state of
+add-ons and will try to sync it with files on disk. As a result, due to races, a single add-on
+can be updated multiple times in a row after upgrading the master. Long-term we should fix this
+by using a similar mechanisms as controller manager or scheduler. However, currently add-on
+manager is just a bash script and adding a master election mechanism would not be easy.
+
+## Adding replica
+
+Command to add new replica:
+
+```
+KUBE_REPLICATE_EXISTING_MASTER=true KUBE_GCE_ZONE=us-central1-b kubernetes/cluster/kube-up.sh
+```
+
+A simplified algorithm for adding a new replica for the master is the following:
+
+```
+1. [only if using load balancer] If there is no load balancer for this cluster:
+  1. Create load balancer using ephemeral IP address
+  2. Add existing apiserver to the load balancer
+  3. Wait until load balancer is working, i.e. all data is propagated, in GCE up to 20 min (sic!)
+  4. Update DNS to point to the load balancer.
+2. Clone existing master (create a new VM with the same configuration) including
+   all env variables (certificates, IP ranges etc), with the exception of
+   `INITIAL_ETCD_CLUSTER`.
+3. SSH to an existing master and run the following command to extend etcd cluster
+   with the new instance:
+   `curl <existing_master>:4001/v2/members -XPOST -H "Content-Type: application/json" -d '{"peerURLs":["http://<new_master>:2380"]}'`
+4. Add IP address of the new apiserver to the load balancer or DNS.
+```
+
+## Deleting replica
+
+Command to delete one replica:
+
+```
+KUBE_DELETE_NODES=false KUBE_GCE_ZONE=us-central1-b kubernetes/cluster/kube-down.sh
+```
+
+A simplified algorithm for deleting an existing replica for the master is the following:
+
+```
+1. Remove replica IP address from the load balancer or DNS configuration
+2. SSH to one of the remaining masters and run the following command to remove replica from the cluster:
+  `curl etcd-0:4001/v2/members/<id> -XDELETE -L`
+3. Delete replica VM
+4. If load balancer has only a single target instance, then delete load balancer
+```
+
+## Upgrades
+
+Upgrading replicated master will be possible by upgrading them one by one using existing tools
```

Comment:
Addon managers of different versions are going to fight during the rolling upgrade.